crypto/src/mls/credential/trust_anchor.rs

```diff
@@ -0,0 +1,1647 @@
+                    anchors
+                        .iter()
+                        .any(|anchor| String::from_utf8_lossy(anchor.domain_name()) == self.domain_name)
+                })
+            {
+                return Err(CryptoError::DuplicateDomainName);
+            }
+        }
+
+        // self signed or incomplete chains are not allowed
+        let leaf_cert = certificate_chain.first().ok_or(CryptoError::InvalidCertificateChain)?;
+
+        // validate domain in the leaf matches with the one supplied
+        let domain_name = extract_domain_name(leaf_cert)?;
+        if domain_name != self.domain_name {
+            return Err(CryptoError::DomainNamesDontMatch);
+        }
+
+        // verify the whole chain
+        let root_cert = certificate_chain
+            .iter()
+            .map(Ok)
+            .reduce(
+                |child, parent| -> Result<&Certificate, openmls_traits::types::CryptoError> {
+                    let child = child?;
+                    let parent = parent?;
+                    child.is_valid()?;
+                    child.is_signed_by(backend.crypto(), parent)?;
+                    Ok(parent)
+                },
+            )
+            .unwrap()
+            .map_err(MlsError::from)?;
+        // ensure that the root is also valid
+        root_cert.is_valid().map_err(MlsError::from)?;
+
+        check_root_in_trust_store(root_cert)?;
+
+        let encoded_chain = certificate_chain
+            .into_iter()
+            .map(|cert| cert.to_der())
+            .collect::<Result<_, x509_cert::der::Error>>()?;
+        Ok(encoded_chain)
+    }
+}
+
+/// Checks the root cert against the trust store. In wasm maybe use webpki-roots (https://github.com/rustls/webpki-roots) crate
+fn check_root_in_trust_store(_root: &Certificate) -> CryptoResult<()> {
+    // TODO: verify root certificate is valid in the device's trust store
+    Ok(())
+}
+
+impl MlsConversation {
+    /// Validates the certificate chain
+    pub(crate) fn validate_received_trust_anchors(
+        old_group_context: &GroupContext,
+        commit_group_context: &GroupContext,
+        backend: &MlsCryptoProvider,
+    ) -> CryptoResult<()> {
+        if let Some(anchors) = commit_group_context.extensions().per_domain_trust_anchors() {
+            // find new anchors
+            anchors
+                .iter()
+                .filter(|anchor| {
+                    old_group_context.extensions().per_domain_trust_anchors().is_none()
+                        || old_group_context
+                            .extensions()
+                            .per_domain_trust_anchors()
+                            .is_some_and(|old_anchors| {
+                                !old_anchors
+                                    .iter()
+                                    .any(|old_anchor| old_anchor.certificate_chain() == anchor.certificate_chain())
+                            })
+                })
```

Comment:
This reads better IMO. Also you could extract this into a `get_trust_anchor_diff()` method

```suggestion
if let Some(new_anchors) = commit_group_context.extensions().per_domain_trust_anchors() {
            // find new anchors
            new_anchors
                .iter()
                .filter(|new_anchor| {
                    let old_trust_anchor = old_group_context.extensions().per_domain_trust_anchors();

                    let no_old_trust_anchor = old_trust_anchor.is_none();
                    let empty_old_trust_anchor = old_trust_anchor.map(|a| a.is_empty()).unwrap_or_default();

                    let anchor_is_new = old_trust_anchor.is_some_and(|old_anchor| {
                        !old_anchor
                            .iter()
                            .any(|old_anchor| old_anchor.certificate_chain() == new_anchor.certificate_chain())
                    });

                    no_old_trust_anchor || empty_old_trust_anchor || anchor_is_new
                })
```